### PR TITLE
Use SimpleURLLoader for url request in RewardsServiceImpl

### PR DESCRIPTION
**Pull request by simonhong**
Fix https://github.com/brave/brave-browser/issues/4277

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [x] Windows
  - [x] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests && npm run test-security`) on
  - [x] Windows
  - [x] macOS
  - [ ] Linux
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection

## Test Plan:

Close https://github.com/brave/brave-browser/issues/4277
## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Comment by simonhong** [comment]
@NejcZdovc With this PR, I got below log when I click rewards button after loading youtube.

Below log is new log with this PR.
```
[97932:775:0501/100214.087064:VERBOSE1:ledger_impl.cc(1126)]
[ RESPONSE - WalletPropertiesCallback ]
> time: 1556672534
> result: Success
> response: {"altcurrency":"BAT","probi":"30000000000000000000","balance":"30.0000","unconfirmed":"0.0000","rates":{"AED":1.3555394475,"ARS":16.3635393309,"AUD":0.523911891,"BAT":1,"BCH":0.0015640192557,"BRL":1.456044768,"BTC":0.0000716988387,"BTG":0.0241333921749,"CAD":0.4969800816,"CHF":0.3759234804,"CNY":2.486966976,"DASH":0.0034750115883,"DKK":2.4618249621,"ETH":0.002386226964112512,"EUR":0.3297430662,"GBP":0.2850646041,"HKD":2.8952285553,"ILS":1.3320322365,"INR":25.761430755,"JPY":41.1040301394,"KES":37.3975481739,"LBA":6.64559697460832,"LTC":0.005420534058,"MXN":7.032161874,"NOK":3.1924600785,"NZD":0.5536003545,"PHP":19.2102330606,"PLN":1.4165437968,"SEK":3.507150411,"SGD":0.5027812332,"USD":0.37261504423776165,"XAG":0.0245931482004,"XAU":0.0002896848597,"XPD":0.0002725692483,"XPT":0.0004076120865,"XRP":1.27935504915},"parameters":{"adFree":{"currency":"BAT","fee":{"BAT":20},"choices":{"BAT":[10,15,20,30,50,100]},"range":{"BAT":[10,100]},"days":30,"nanos":1}},"grants":[{"altcurrency":"BAT","expiryTime":1640995200,"probi":"30000000000000000000","type":"ugp"}]}
> headers cache-control: private
> headers connection: keep-alive
> headers content-encoding: gzip
> headers content-type: application/json; charset=utf-8
> headers date: Wed, 01 May 2019 01:02:14 GMT
> headers server: Cowboy
> headers transfer-encoding: chunked
> headers vary: accept-encoding
> headers via: 1.1 vegur
> headers x-rate-limit-limit: 60
> headers x-rate-limit-remaining: 59
> headers x-rate-limit-reset: 1556672594
[ END RESPONSE ]
[97932:775:0501/100214.151614:VERBOSE1:ledger_impl.cc(1126)]
[ RESPONSE - getGrantsCallback ]
> time: 1556672534
> result: Failure
> response:
> headers cache-control: no-cache
> headers connection: keep-alive
> headers content-encoding: gzip
> headers content-type: application/json; charset=utf-8
> headers date: Wed, 01 May 2019 01:02:14 GMT
> headers server: Cowboy
> headers transfer-encoding: chunked
> headers vary: accept-encoding
> headers via: 1.1 vegur
> headers x-rate-limit-limit: 60
> headers x-rate-limit-remaining: 58
> headers x-rate-limit-reset: 1556672570
[ END RESPONSE ]
[97932:775:0501/100214.151747:WARNING:ledger_impl.cc(956)] Failed to refresh grant, will try again in 397
```

Below log is by w/o this PR.
```
[98089:775:0501/100715.158865:VERBOSE1:ledger_impl.cc(1126)]
[ RESPONSE - WalletPropertiesCallback ]
> time: 1556672835
> result: Success
> response: {"altcurrency":"BAT","probi":"30000000000000000000","balance":"30.0000","unconfirmed":"0.0000","rates":{"AED":1.3555394475,"ARS":16.3635393309,"AUD":0.523911891,"BAT":1,"BCH":0.0015640192557,"BRL":1.456044768,"BTC":0.0000716988387,"BTG":0.0241333921749,"CAD":0.4969800816,"CHF":0.3759234804,"CNY":2.486966976,"DASH":0.0034750115883,"DKK":2.4618249621,"ETH":0.002386226964112512,"EUR":0.3297430662,"GBP":0.2850646041,"HKD":2.8952285553,"ILS":1.3320322365,"INR":25.761430755,"JPY":41.1040301394,"KES":37.3975481739,"LBA":6.64559697460832,"LTC":0.005420534058,"MXN":7.032161874,"NOK":3.1924600785,"NZD":0.5536003545,"PHP":19.2102330606,"PLN":1.4165437968,"SEK":3.507150411,"SGD":0.5027812332,"USD":0.37261504423776165,"XAG":0.0245931482004,"XAU":0.0002896848597,"XPD":0.0002725692483,"XPT":0.0004076120865,"XRP":1.27935504915},"parameters":{"adFree":{"currency":"BAT","fee":{"BAT":20},"choices":{"BAT":[10,15,20,30,50,100]},"range":{"BAT":[10,100]},"days":30,"nanos":1}},"grants":[{"altcurrency":"BAT","expiryTime":1640995200,"probi":"30000000000000000000","type":"ugp"}]}
> headers cache-control: private
> headers connection: keep-alive
> headers content-encoding: gzip
> headers content-type: application/json; charset=utf-8
> headers date: Wed, 01 May 2019 01:07:15 GMT
> headers server: Cowboy
> headers transfer-encoding: chunked
> headers vary: accept-encoding
> headers via: 1.1 vegur
> headers x-rate-limit-limit: 60
> headers x-rate-limit-remaining: 59
> headers x-rate-limit-reset: 1556672895
[ END RESPONSE ]
[98089:775:0501/100715.352723:VERBOSE1:ledger_impl.cc(1126)]
[ RESPONSE - getGrantsCallback ]
> time: 1556672835
> result: Failure
> response: {"statusCode":404,"error":"Not Found","message":"promotion not available"}
> headers cache-control: no-cache
> headers connection: keep-alive
> headers content-encoding: gzip
> headers content-type: application/json; charset=utf-8
> headers date: Wed, 01 May 2019 01:07:15 GMT
> headers server: Cowboy
> headers transfer-encoding: chunked
> headers vary: accept-encoding
> headers via: 1.1 vegur
> headers x-rate-limit-limit: 60
> headers x-rate-limit-remaining: 59
> headers x-rate-limit-reset: 1556672895
[ END RESPONSE ]
```
With this PR, exsiting rewards service browser tests are failed.
Can you check this PR causes regression or existing test should be modified?

**Comment by simonhong** [comment]
With `SimpleURLLoader::SetAllowHttpErrorResults(true)`,
I got same response as before. but, still many tests are failed.

File: components/brave_rewards/browser/rewards_service_browsertest.cc [hunk 1]
```diff
@@ -45,6 +43,28 @@ std::unique_ptr<net::test_server::HttpResponse> HandleRequest(
...
+}
+
+void Split(std::vector<std::string>* tmp,
+           const std::string& query,
+           char delimiter) {
+  DCHECK(tmp);
+  std::stringstream ss(query);
+  std::string item;
+  while (std::getline(ss, item, delimiter)) {
+    if (query[0] != '\n') {
```
Comment (jasonrsadler) [inline components/brave_rewards/browser/rewards_service_browsertest.cc]: Would this cause the CRLF problems on windows?

**Comment by simonhong** [comment]
All tests are passed. I think this PR is ready to review.

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 2]
```diff
@@ -1236,33 +1238,56 @@ void RewardsServiceImpl::LoadURL(
...
-
+  if (test_response_callback_) {
+    std::string test_response;
+    std::map<std::string, std::string> test_headers;
+    test_response_callback_.Run(url, &test_response, &test_headers),
+    callback(200, test_response, test_headers);
+    return;
+  }
+
+  net::NetworkTrafficAnnotationTag traffic_annotation =
```
Comment (simonhong) [inline components/brave_rewards/browser/rewards_service_impl.cc]: @NejcZdovc Need help to fill this annotation. (I don't have any background about rewards service.)

File: components/brave_rewards/browser/rewards_service_browsertest.cc (hunk 1 above)
Comment (simonhong) [reply components/brave_rewards/browser/rewards_service_browsertest.cc]: Well, I didn't test this PR on Windows yet. (I didn't modify this logic in this PR)
Will check this PR on Windows.

Comment (jasonrsadler) [reply components/brave_rewards/browser/rewards_service_browsertest.cc]: 👍 May not be relevant here but a couple of us have been bitten by CRLF delimiters.

Comment (simonhong) [reply components/brave_rewards/browser/rewards_service_browsertest.cc]: Rewards browser test works well on my local Win10. :)

Comment (iefremov) [reply components/brave_rewards/browser/rewards_service_browsertest.cc]: maybe just `base::SplitString`?

File: components/brave_rewards/browser/rewards_service_impl.cc (hunk 2 above)
Comment (iefremov) [inline components/brave_rewards/browser/rewards_service_impl.cc]: sometimes its better to extract annotation from the function, just to improve readability (see many `GetNetworkTrafficAnnotationTag()` around)

**Comment by iefremov** [comment]
LGTM

File: components/brave_rewards/browser/rewards_service_browsertest.cc (hunk 1 above)
Comment (simonhong) [reply components/brave_rewards/browser/rewards_service_browsertest.cc]: Good to know this api! thanks. Done.

File: components/brave_rewards/browser/rewards_service_impl.cc (hunk 2 above)
Comment (simonhong) [reply components/brave_rewards/browser/rewards_service_impl.cc]: Done.

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 3]
```diff
@@ -302,6 +304,54 @@ void EnsureRewardsBaseDirectoryExists(const base::FilePath& path) {
...
+          "User-initiated for direct tipping or on a set interval while Brave "
+          "is running for monthly contributions."
+        data:
+          "Publisher and contribution data."
+        destination: WEBSITE
+      }
+      policy {
+        cookies_allowed: NO
+        setting:
+          "You can enable or disable this feature via the Brave Shields icon "
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/rewards_service_impl.cc]: `You can enable or disable this feature via the Brave Shields icon` -> `You can enable or disable this feature via the BAT icon in the URL bar`

Comment (simonhong) [reply components/brave_rewards/browser/rewards_service_impl.cc]: Done. thanks!

**Pull request merged**